### PR TITLE
`torch` lazy loading

### PR DESCRIPTION
**Pull request by AdrianSosic**
This PR finalizes the import refactoring such that `torch` (and the dependent packages `gpytorch`/`botorch`) are loaded lazily for improved package import speed.

Further, new import tests ensure that:
* all BayBE modules can be loaded without errors
* `torch` is not eagerly loaded (to prevent regressions)

### CODE REVIEW

**Comment by AdrianSosic** [comment]
@AVHopp, @Scienfitz, @rjavadi: this PR finally concludes the torch refactoring. Note that it sits on top of another PR, hence the extended diff ...

**Comment by AdrianSosic** [comment]
Perhaps one additional comment: with this PR the startup time is already significantly reduced and we got rid of the (by far) biggest chunk. The next "block" would be `sklearn`, which can be in principle removed in the same way. However, I think we should postpone this and first focus on more urgent issues. Thoughts?

File: CHANGELOG.md [hunk 1]
```diff
@@ -15,12 +15,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `hypothesis` strategies and roundtrip test for kernels and objectives
 
 ### Changed
-- `torch` numeric types are now loaded lazily
 - Reorganized acquisition.py into `acquisition` subpackage
-- `torch` is imported lazily in `surrogates`
 - Acquisition functions are now their own objects
 - `acquisition_function_cls` constructor parameter renamed to `acquisition_function`
+- `torch`, `gpytorch` and `botorch` are lazy-loaded for improved startup time
```
Comment (AdrianSosic) [inline CHANGELOG.md:21]: (continued): and lazy-loading `sklearn` would also have the benefit that we could really turn in into an optional dependency ...

File: CHANGELOG.md [hunk 2]
```diff
@@ -15,12 +15,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
+### Removed
```
Comment (AVHopp) [inline CHANGELOG.md:23]: @AdrianSosic will review after the preceding PR was merged. Would appreciate a quick ping once this is the case :)

Comment (AdrianSosic) [reply CHANGELOG.md:23]: Sure, thanks. "Ping" btw :D

File: tests/test_imports.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
+"""Import tests."""
```
Comment (AVHopp) [inline tests/test_imports.py:1]: I'd rephrase that since that could also be understood as "We import tests here", while it means "We test imports here".

File: tests/test_imports.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+import importlib
+import os
+import pkgutil
+import subprocess
+import sys
+from collections.abc import Sequence
+
+import pytest
+from pytest import param
```
Comment (AVHopp) [inline tests/test_imports.py:11]: Is there any reason for this additional import and not just calling `pytest.param`?

File: tests/test_imports.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+def make_import_check(modules: Sequence[str], target: str) -> str:
+    """Create code that tests if importing the given modules also imports the target.
+
+    Args:
+        modules: The modules to be imported by the created code.
+        target: The target module whose presence is to be checked after the import.
+
+    Returns:
+        str: Code that signals the presence of the target via a non-zero exit code.
```
Comment (AVHopp) [inline tests/test_imports.py:40]: The explicit type hint is added automatically, please remove.

File: tox.ini [hunk 6]
```diff
@@ -9,6 +9,7 @@ extras = test,chem,examples,simulation,onnx
 passenv = CI
 setenv =
     SMOKE_TEST = true
+    BAYBE_TEST_ENV = FULLTEST
```
Comment (AVHopp) [inline tox.ini:12]: Why are the additions in this file necessary?

**Review by AVHopp — APPROVED** [review]
Looking good, just some minor questions.

File: tests/test_imports.py (hunk 5 above)
Comment (AdrianSosic) [reply tests/test_imports.py]: Thanks, that part was auto-generated and I forgot to delete it! Done

File: tests/test_imports.py (hunk 3 above)
Comment (AdrianSosic) [reply tests/test_imports.py]: I see what you mean. Changed it 👍🏼

File: tests/test_imports.py (hunk 4 above)
Comment (AdrianSosic) [reply tests/test_imports.py:11]: In this file it doesn't make a huge difference but I usually tend to import this separately to keep the parametrization code short. Have look, for instance, into `test_parameter_validation.py` and you'll see why 😄 

Since it pretty much comes at zero additional costs, there is no harm to also keep it here, so unless you really want me to remove it, I'd say let's just keep it for better readability

File: tox.ini (hunk 6 above)
Comment (AdrianSosic) [reply tox.ini:12]: The reason is because the import tests will only work in a `dev` environment, since otherwise packages like `xyzpy` are not available and will trigger errors that should not be part of the test. See beginning of `test_imports.py`:
```python
pytestmark = pytest.mark.skipif(
    os.environ.get("BAYBE_TEST_ENV") != "FULLTEST",
    reason="Only possible in FULLTEST environment.",
)
```

I already foresee a smarter way here which will also make all our little helpers like `_ONNX_INSTALLED` and the various try-except blocks obsolete, and which will then enable to run the tests also in a basic environment. But that belongs to another PR, where I can then theoretically remove these variables again. However, I think they also might become handy in other situations. Anyway, for now they are needed (unless you have a better idea?) and we can decide later what to do with them 👍🏼

File: CHANGELOG.md (hunk 1 above)
Comment (Scienfitz) [reply CHANGELOG.md:21]: sure but i think thats low prio
curious: how much loading time has sklearn vs torch anyway?

Comment (AdrianSosic) [reply CHANGELOG.md:23]: Yes, agree, no high prio...

For the numbers: 
* before: loading botorch roughly took 1.2s in the import tree
* after: sklearn takes rougly 0.5s

Feel free to resolve

File: tests/test_imports.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+import os
+import pkgutil
+import subprocess
+import sys
+from collections.abc import Sequence
+
+import pytest
+from pytest import param
+
+pytestmark = pytest.mark.skipif(
```
Comment (Scienfitz) [inline tests/test_imports.py:13]: this is not used anywhere or? Id guess the intent is to use it like a `@pytestmark` before the tests below, or is this just here to skip the entire file here somehow?

Comment (AdrianSosic) [reply tests/test_imports.py:13]: This is in fact `pytest` magic to indicate the latter: https://docs.pytest.org/en/7.1.x/example/markers.html#marking-whole-classes-or-modules

As pointed out to Alex, at some appropriate point I want to improve our optional dependency loading a bit and then this mark might become irrelevant, but for now I think we need it. Please resolve if OK

File: tests/test_imports.py [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    os.environ.get("BAYBE_TEST_ENV") != "FULLTEST",
+    reason="Only possible in FULLTEST environment.",
+)
+
+_EAGER_LOADING_EXIT_CODE = 42
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply tests/test_imports.py:18]: Believe it or not, 1337 is what I tried first but then I learned that the exit codes need to be in 0...255, with 0...127 for custom codes and 127...255 for system codes 😄 

I added it to make sure we are not actually getting another (unexpected) nonzero code instead. This is similar to testing for specific `ValueErrors` in tests but then getting a `ValueError` cause by some other unexpected problem, which would make your test succeed incorrectly (that's something that I still need to change in the validation tests, btw). 

Overall, I'm not super happy with the entire subprocess approach I used here (too much string hassle) but it does the job and I couldn't come up with a better idea. Do you perhaps have one?

**Review by Scienfitz — APPROVED** [review]
